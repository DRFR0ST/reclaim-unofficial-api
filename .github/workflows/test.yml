name: ðŸ§ª Run Tests

on:
  push:
    branches: [ main ]
  pull_request:
    branches: [ main ]
  schedule:
    - cron: '0 0 * * *'  # Run every day at midnight to ensure the reverse engineered API is up to date.

jobs:
  test:
    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v3
    - uses: oven-sh/setup-bun@v1

    - name: Install dependencies
      run: bun install

    - name: Run tests
      uses: nick-fields/retry@v2
      with:
        timeout_minutes: 5
        max_attempts: 3
        command: bun run test
      env:
        RECLAIM_ACCESS_TOKEN: ${{ secrets.RECLAIM_ACCESS_TOKEN }}
    
    - name: Build library
      run: bun run build
      env:
        RECLAIM_ACCESS_TOKEN: ${{ secrets.RECLAIM_ACCESS_TOKEN }}