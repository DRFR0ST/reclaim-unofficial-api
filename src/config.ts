
const RECLAIM_API_URL = "https://api.app.reclaim.ai";
const RECLAIM_ACCESS_TOKEN = process?.env?.RECLAIM_ACCESS_TOKEN;
const RECLAIM_VERBOSE = process?.env?.RECLAIM_VERBOSE === "true";

/**
 * @description Configuration for the Reclaim API client.
 */
export const config = Object.freeze({
    reclaim: {
        apiUrl: RECLAIM_API_URL,
        accessToken: RECLAIM_ACCESS_TOKEN
    },
    verbose: RECLAIM_VERBOSE
});