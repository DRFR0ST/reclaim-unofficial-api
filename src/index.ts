export { ReclaimClient } from "./client"
export type { ReclaimUser, ReclaimCalendar, ReclaimTask, ReclaimEndpoints } from './interfaces';