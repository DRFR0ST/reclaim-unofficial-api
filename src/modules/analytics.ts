import {
  ReclaimEndpoints,
  ReclaimAnalytics as ReclaimAnalyticsType,
  ReclaimAnalyticsGetPayload,
  ReclaimAnalyticsMetricName,
} from "../types";
import { ReclaimClient } from "../client";

/**
 * A class for interacting with Reclaim analytics.
 */
export class ReclaimAnalytics {
  private path = ReclaimEndpoints.Analytics;
  private client: ReclaimClient;

  constructor(client: ReclaimClient) {
    this.client = client;
  }

  /**
   * @description The default metric name to use when getting analytics.
   */
  private defaultMetricName: ReclaimAnalyticsMetricName[] = [
    "DURATION_BY_CATEGORY",
    "DURATION_BY_DATE_BY_CATEGORY",
    "FOCUS_TIME_DURATION_BY_DAY_OF_WEEK",
    "MEETING_CALLOUT_GROUP",
    "MEETING_TIME_DURATION_BY_DAY_OF_WEEK",
    "PERSONAL_HABITS_AND_TASKS_DURATION_BY_DATE_BY_NAME",
    "ROI_CALLOUT_GROUP",
    "TIME_WITH_PEOPLE_DURATION_BY_PERSON",
    "WORK_HABITS_AND_TASKS_DURATION_BY_DATE_BY_NAME",
    "WORK_LIFE_CALLOUT_GROUP",
  ];

  /**
   * @description Get the user's analytics.
   * @returns {ReclaimAnalytics}
   */
  async get({
    start,
    end,
    metricName,
    groupingInterval,
  }: ReclaimAnalyticsGetPayload = {}): Promise<ReclaimAnalyticsType> {
    // Set the start and end dates to the last 7 days and the next 7 days, respectively. Format YYYY-MM-DD.
    if (!start)
      start = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000)
        .toISOString()
        .split("T")[0];
    if (!end)
      end = new Date(Date.now() + 7 * 24 * 60 * 60 * 1000)
        .toISOString()
        .split("T")[0];
    if (!metricName) metricName = this.defaultMetricName;
    if (!groupingInterval) groupingInterval = "DAILY";

    // Ensure that metricName is an array of strings
    if (!Array.isArray(metricName))
      throw new Error("The property metricName must be an array of strings.");

    // Compose the URL.
    const url = `${
      this.path
    }?start=${start}&end=${end}&metricName=${metricName.join(
      "%2C"
    )}&groupingInterval=${groupingInterval}`;

    return await this.client._fetcher(url, {
      method: "GET",
    });
  }
}
