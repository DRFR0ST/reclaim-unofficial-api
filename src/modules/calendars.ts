import {
  ReclaimCalendar,
  ReclaimEndpoints,
} from "../types";
import { ReclaimClient } from "../client";

/**
 * A class for interacting with Reclaim calendars.
 */
export class ReclaimCalendars {
  private path = ReclaimEndpoints.Calendars;
  private client: ReclaimClient;

  constructor(client: ReclaimClient) {
    this.client = client;
  }

  /**
   * @description Get the primary calendar.
   * @returns {ReclaimUser}
   */
  async primary(): Promise<ReclaimCalendar> {
    return await this.client._fetcher(`${this.path}/primary`, {
      method: "GET",
    });
  }
}
