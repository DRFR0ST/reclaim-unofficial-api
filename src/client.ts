import { ReclaimTasks } from "./modules/tasks";
import { ReclaimUsers } from "./modules/users";
import { ReclaimCalendars } from "./modules/calendars";
import { config } from "./config";

export class ReclaimClient {
  private accessToken = "";

  constructor(options?: { accessToken: string }) {
    if (!options?.accessToken && !config.reclaim.accessToken)
      throw new Error("Reclaim access token is required.");

    this.accessToken = options?.accessToken || config.reclaim.accessToken!;
  }

  get tasks() {
    return new ReclaimTasks(this);
  }

  get users() {
    return new ReclaimUsers(this);
  }

  get calendars() {
    return new ReclaimCalendars(this);
  }

  async _fetcher(endpoint: string, options: RequestInit) {
    const opts = {
      ...options,
      headers: {
        ...options.headers,
        Authorization: `Bearer ${this.accessToken}`,
      },
    };

    try {
      const response = await fetch(
        `${config.reclaim.apiUrl}/api/${endpoint}`,
        opts
      );
      const json = await response.json();

      return json;
    } catch (err) {
      console.error(err);
      throw new Error("Error fetching data from Reclaim API.");
    }
  }
}
