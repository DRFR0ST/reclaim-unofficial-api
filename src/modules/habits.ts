import {
  ReclaimEndpoints,
  ReclaimHabit,
  ReclaimHabitCreate,
} from "../types";
import { ReclaimClient } from "../client";
import { filterArray } from "../utils";

/**
 * A class for interacting with Reclaim habits.
 */
export class ReclaimHabits {
  private path = ReclaimEndpoints.Habits;
  private client: ReclaimClient;

  constructor(client: ReclaimClient) {
    this.client = client;
  }

  /**
   * @description Create a new habit.
   * @param habit The habit to create.
   * @returns {ReclaimHabit}
   */
  async create(habit: ReclaimHabitCreate): Promise<ReclaimHabit> {
    const request = await this.client._fetcher(this.path, {
      method: "POST",
      body: JSON.stringify(habit),
    });

    const createdHabit = request.find((h: ReclaimHabit) => h.title === habit.title);

    return createdHabit;
  }

  async search(filters?: Partial<ReclaimHabit>): Promise<ReclaimHabit[]> {
    const response = (await this.client._fetcher(this.path, {
      method: "GET",
    })) as ReclaimHabit[];

    if (!filters) return response;

    // Filter the response based on the filter object
    return filterArray(response, filters);
  }

  /**
   * @description Get a single habit by ID.
   * @param id The ID of the habit to get.
   * @returns {ReclaimHabit | null}
   */
  async get(id: number): Promise<ReclaimHabit | null> {
    const habits = await this.search();

    return habits.find((habit) => habit.id === id) || null;
  }

  /**
   * @description Update a habit.
   * @param habit The habit to update.
   * @returns {ReclaimHabit}
   */
  async update(id: number, habit: Partial<ReclaimHabitCreate>): Promise<ReclaimHabit> {
    return await this.client._fetcher(`${this.path}/${id}`, {
      method: "PATCH",
      body: JSON.stringify(habit),
    });
  }

  /**
   * @description Delete a habit.
   * @param id The ID of the habit to delete.
   * @returns {null}
   */
  async delete(id: number): Promise<{ taskOrHabit: ReclaimHabit, events: unknown[] }> {
    return await this.client._fetcher(`planner/policy/habit/${id}`, {
      method: "DELETE",
    });
  }
}
