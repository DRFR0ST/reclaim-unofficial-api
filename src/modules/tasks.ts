import {
  ReclaimEndpoints,
  ReclaimTask,
  ReclaimTaskCreate,
} from "../types";
import { ReclaimClient } from "../client";
import { filterArray } from "../utils";

/**
 * A class for interacting with Reclaim tasks.
 */
export class ReclaimTasks {
  private path = ReclaimEndpoints.Tasks;
  private client: ReclaimClient;

  constructor(client: ReclaimClient) {
    this.client = client;
  }

  /**
   * @description Get all tasks, optionally filtered by a filter object.
   * @param filters The filters to apply.
   * @returns {ReclaimTask[]}
   */
  async search(filters?: Partial<ReclaimTask>): Promise<ReclaimTask[]> {
    const response = (await this.client._fetcher(`${this.path}`, {
      method: "GET",
    })) as ReclaimTask[];

    if (!filters) return response;

    // Filter the response based on the filter object
    return filterArray(response, filters);
  }

  /**
   * @description Get a single task by ID.
   * @param id The ID of the task to get.
   * @returns {ReclaimTask}
   */
  async get(id: number): Promise<ReclaimTask> {
    return await this.client._fetcher(`${this.path}/${id}`, {
      method: "GET",
    });
  }

  /**
   * @description Create a new task.
   * @param task The task to create.
   * @returns {ReclaimTask}
   */
  async create(task: ReclaimTaskCreate): Promise<ReclaimTask> {
    return await this.client._fetcher(this.path, {
      method: "POST",
      body: JSON.stringify(task),
    });
  }

  /**
   * @description Update a task.
   * @param id The ID of the task to update.
   * @param task The task to update.
   * @returns {ReclaimTask}
   */
  async update(id: number, task: Partial<ReclaimTask>): Promise<ReclaimTask> {
    return await this.client._fetcher(`${this.path}/${id}`, {
      method: "PATCH",
      body: JSON.stringify(task),
    });
  }

  /**
   * @description Delete a task.
   * @param id The ID of the task to delete.
   * @returns {null}
   */
  async delete(id: number): Promise<null> {
    return await this.client._fetcher(`${this.path}/${id}`, {
      method: "DELETE",
    });
  }
}
