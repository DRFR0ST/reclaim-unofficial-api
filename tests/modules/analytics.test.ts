import { beforeEach, describe, expect, test } from "bun:test";
import { ReclaimClient } from "../../src";
import { ReclaimAnalytics } from "../../src/modules/analytics";

describe("ReclaimAnalytics", () => {
  let analytics: ReclaimAnalytics;

  beforeEach(() => {
    const client = new ReclaimClient();
    analytics = client.analytics;
  });

  test("should be created", () => {
    expect(analytics).toBeTruthy();
  });

  test("should get the user's analytics", async () => {
    const results = await analytics.get();

    expect(results.result.metrics).toBeTruthy();
    expect(results.result.metrics).toBeArray();
  });

  test("should get the user's analytics for a specific timespan", async () => {
    const results = await analytics.get({
      start: "2023-11-01",
      end: "2023-11-30",
      metricName: ["DURATION_BY_CATEGORY"],
    });

    expect(results.result.metrics).toBeTruthy();
    expect(results.result.metrics).toBeArray();
    expect(results.result.metrics).toMatchSnapshot();
  });
});
