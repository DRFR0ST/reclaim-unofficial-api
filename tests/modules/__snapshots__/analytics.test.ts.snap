// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`ReclaimAnalytics should get the user's analytics for a specific timespan 1`] = `
[
  {
    "metric": {
      "@type": "AnalyticsDurationByCategoryMetric",
      "appendixKeys": {},
      "bucket": "ANALYTICS_CATEGORY",
      "description": "Duration by category",
      "noData": false,
      "type": "durationByCategory",
      "unit": "DURATION_MINUTES",
      "values": [
        {
          "key": "TEAM_MEETING",
          "value": 0,
        },
        {
          "key": "EXTERNAL_MEETING",
          "value": 17,
        },
        {
          "key": "ONE_ON_ONE",
          "value": 14,
        },
        {
          "key": "BREAK",
          "value": 5,
        },
        {
          "key": "DEEP_WORK",
          "value": 62,
        },
        {
          "key": "SHALLOW_WORK",
          "value": 30,
        },
      ],
    },
    "name": "DURATION_BY_CATEGORY",
  },
]
`;
